src/kdf.cpp

```diff
@@ -0,0 +1,142 @@
+/*
+ * #%L
+ * %%
+ * Copyright (C) 2020 BMW Car IT GmbH
+ * %%
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * #L%
+ */
+#include "mococrw/kdf.h"
+#include "mococrw/openssl_wrap.h"
+
+namespace mococrw {
+
+KeyDerivationFunction::~KeyDerivationFunction() = default;
+
+
+
+
+class PBKDF2::Impl
+{
+public:
+    Impl(openssl::DigestTypes hashFunction, uint32_t iterations):
+        _hashFunction(hashFunction), _iterations(iterations) {}
+
+    ~Impl() = default;
+
+    Impl(Impl&&) = default;
+
+    Impl(Impl&) = default;
+
+    std::vector<uint8_t> deriveKey(const std::vector<uint8_t> &password,
+                                   const size_t outputLength,
+                                   const std::vector<uint8_t> &salt)
+    {
+        std::vector<uint8_t> derivedKey(outputLength);
+        const EVP_MD* digestFn = _getMDPtrFromDigestType(_hashFunction);
+        openssl::_PKCS5_PBKDF2_HMAC(reinterpret_cast<const char*>(password.data()), password.size(),
+                                    salt.data(), salt.size(),
+                                    _iterations, digestFn, outputLength, derivedKey.data());
+
+        return derivedKey;
+    }
+
+private:
+    openssl::DigestTypes _hashFunction;
+    uint32_t _iterations;
+};
+
+PBKDF2::PBKDF2(openssl::DigestTypes hashFunction, uint32_t iterations)
+{
+    _impl = std::make_unique<PBKDF2::Impl>(hashFunction, iterations);
+}
+
+PBKDF2::~PBKDF2() = default;
+
+std::vector<uint8_t> PBKDF2::deriveKey(const std::vector<uint8_t> &password,
+                                       const size_t outputLength,
+                                       const std::vector<uint8_t> &salt)
+{
+    return _impl->deriveKey(password, outputLength, salt);
+}
+
+PBKDF2::PBKDF2(PBKDF2&& other) = default;
+
+PBKDF2& PBKDF2::operator=(PBKDF2&& other) {
+    this->_impl = std::move(other._impl);
+    return *this;
+}
+
+PBKDF2::PBKDF2(const PBKDF2& other)
+    : _impl(std::make_unique<PBKDF2::Impl>(*(other._impl))) {}
```

Comment:
Noise: The brackets around other._impl should not be needed. The dot binds stronger than *

Reply:
Done